### PR TITLE
KAFKA-5845; KafkaController should send LeaderAndIsrRequest to broker which starts very soon after shutdown

### PR DESCRIPTION
**Pull request by lindong28**

### CODE REVIEW

File: core/src/main/scala/kafka/controller/KafkaController.scala [hunk 1]
```diff
@@ -1209,8 +1209,11 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, time: Time, met
       // the odds of processing recent broker changes in a single ControllerEvent (KAFKA-5502).
       val curBrokers = zkUtils.getAllBrokersInCluster().toSet
       val curBrokerIds = curBrokers.map(_.id)
+      val brokerIdsOnEventReceived = brokersOnEventReceived.map(_.toInt).toSet
       val liveOrShuttingDownBrokerIds = controllerContext.liveOrShuttingDownBrokerIds
-      val newBrokerIds = curBrokerIds -- liveOrShuttingDownBrokerIds
+      // Also call onBrokerStartup for brokers in (curBrokerIds -- brokerIdsOnEventReceived) to take care of the
+      // scenario that the broker has started after the event is received but before the event is processed
+      val newBrokerIds = (curBrokerIds -- liveOrShuttingDownBrokerIds) ++ (curBrokerIds -- brokerIdsOnEventReceived)
```
Comment (tedyu) [inline core/src/main/scala/kafka/controller/KafkaController.scala:1216]: I wonder if the following expression suffices:
```
(curBrokerIds -- liveOrShuttingDownBrokerIds) -- brokerIdsOnEventReceived
```

Comment (lindong28) [reply core/src/main/scala/kafka/controller/KafkaController.scala:1216]: @tedyu Thanks for the comment. I will improve it as you suggested.

**Comment by ijuma** [comment]
cc @onurkaraman @junrao

File: core/src/main/scala/kafka/controller/KafkaController.scala (hunk 1 above)
Comment (lindong28) [reply core/src/main/scala/kafka/controller/KafkaController.scala:1216]: @tedyu I just double checked it again. Your suggestion would not work. The counter example is curBrokerIds=1,2,3. liveOrShuttingdownBrokerId=1,2,3 brokerIdsOnEventReceived=1,2

Comment (lindong28) [reply core/src/main/scala/kafka/controller/KafkaController.scala:1216]: And this is why I originally wrote it in that way.

**Comment by junrao** [comment]
@lindong28 : As James mentioned in the jira, this issue in general is the same as KAFKA-1120. The fix that you provided is not going to solve the problem completely since from the time a watcher is triggered to the time that the controller reads the broker registration, the broker could have been restarted and registered again. In KAFKA-1120, we discussed a more general solution by keeping track the broker epoch (e.g. using ZK session id) in the controller. This way, even if the same broker is de-registered and re-registered quickly, the controller will notice that the broker epoch has changed and can act accordingly.

**Comment by lindong28** [comment]
@junrao Ah I just noticed the reference to KIP-1120 in the JIRA. I agree a broker epoch is more general and is a better solution in terms of design. But I am not sure why the solution in this patch would not solve the problem completely.

I was assuming that regardless of how fast the broker de-register and register itself, the BrokerChangeListener in the controller will always two callbacks. The first callback will provide a list of broker ids without that broker, the second callback will provide a list of brokerids with that broker. If this is not the case, the fix in this patch is not guaranteed to work.

If this assumption is correct, I think the fix in this patch is guaranteed to work. The idea is that the first `brokerIdsOnEventReceived` will not include the broker id and the second `brokerIdsOnEventReceived` will include the broker id. Depending on whether broker register itself before or after controller handles the first broker change event, controller will call onBrokerStartup() for this broker in either the first or the second callback. Does this make sense?

**Comment by junrao** [comment]
@lindong28 : Right, the issue is that the assumption that every event will be notified through ZK watcher is incorrect. The way that ZK watcher works is that it's a one-time watcher. A ZK client registers a child change watcher by doing a child read. When the child changes, the watcher will be fired and the watcher is no longer registered. The firing of the watcher only tells the ZK client that sth has changed, but not what has changed. The ZK client has to do a child read to figure out what has changed (at which point, the watcher can be registered again). However, between the firing of the watcher and the reading of the child value, additional child changes could have happened. The application has to deal with this.

For more details, you can see https://zookeeper.apache.org/doc/r3.4.9/zookeeperProgrammers.html#sc_WatchRememberThese

**Comment by lindong28** [comment]
@junrao Thanks for the reference. I agree. I will close the JIRA.

**Comment by lindong28** [comment]
@junrao

**Comment by lindong28** [comment]
@junrao It just came to me that this patch may be solving a slightly different problem from KAFKA-1120. This patch is solving the regression issue introduced by the single controller patch, where the  controller will not send LeaderAndIsrRequest to a broker even if controler's zk callback did not miss any event. Prior to the single controller patch, the issue may happen if controller's zk callback miss some event, but the chance of that happening is probably much lower than it is now.

If we can solve the KAFKA-1120 using e.g. ZK session ID, that will be great and solves the problem completely. I am just wondering if KAFKA-1120 is addressed after 1.0 releasee, would it be useful to have this patch for 1.0 release?

**Comment by junrao** [comment]
@lindong28 : Before the single threaded controller patch, a broker restart event can be processed in parallel with the broker restart. After the patch, these two events will be processed sequentially in order, which is an improvement. The patch does make it more possible for a broker restart event to be missed if the controlled shutdown takes long. However, this can be avoided by increasing the controller request timeout.

As for this patch, intuitively, we could either use the broker registration when the event is triggered or when the event is processed. However, using both at the same time seems unintuitive and a bit hacky.

We probably won't have time to address KAFKA-1120 in 1.0.0. However, Onur is working on KAFKA-5642 for 1.0.0. This will make the controlled shutdown faster and should help the issue you described here as well.

**Comment by lindong28** [comment]
@junrao I see. Thanks much for your detailed explanation.

**Pull request closed without merging**